### PR TITLE
test/e2e: Check for AM pods and cluster ready & no parallel

### PR DESCRIPTION
**Pull request by paulfantom**
Trying to find a solution to flaky AM test discussed in #2534 

/cc @mxinden

### CODE REVIEW

**Comment by mxinden** [comment]
> `alertmanager_test.go:531: failed to wait for initialized alertmanager mesh: the server rejected our request for an unknown reason (get pods alertmanager-rolling-deploy-0): <nil>`

We need to wait for all Alertmanager pods to become ready before doing any API calls. This should be fixed with the following patch. What do you think @paulfantom?

```diff
diff --git a/test/e2e/alertmanager_test.go b/test/e2e/alertmanager_test.go
index 3519fc56..2d7f11f8 100644
--- a/test/e2e/alertmanager_test.go
+++ b/test/e2e/alertmanager_test.go
@@ -520,16 +520,14 @@ inhibit_rules:
 	if _, err := framework.KubeClient.CoreV1().Secrets(ns).Create(amcfg); err != nil {
 		t.Fatal(err)
 	}
+
 	alertmanager, err = framework.MonClientV1.Alertmanagers(ns).Create(alertmanager)
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	for i := 0; i < int(*alertmanager.Spec.Replicas); i++ {
-		name := "alertmanager-rolling-deploy-" + strconv.Itoa(i)
-		if err := framework.WaitForAlertmanagerInitializedMesh(ns, name, int(*alertmanager.Spec.Replicas)); err != nil {
-			t.Fatal(err)
-		}
+	if err := framework.WaitForAlertmanagerClusterReady(ns, alertmanager.Name, int(*alertmanager.Spec.Replicas)); err != nil {
+		t.Fatal(err)
 	}
 
 	if _, err := testFramework.CreateServiceAndWaitUntilReady(framework.KubeClient, ns, amsvc); err != nil {
@@ -615,6 +613,10 @@ inhibit_rules:
 		t.Fatal(err)
 	}
 
+	if err := framework.WaitForAlertmanagerClusterReady(ns, alertmanager.Name, int(*alertmanager.Spec.Replicas)); err != nil {
+		t.Fatal(err)
+	}
+
 	time.Sleep(time.Minute)
 
 	logs, err = testFramework.GetLogs(framework.KubeClient, ns, podName, "webhook-server")
diff --git a/test/framework/alertmanager.go b/test/framework/alertmanager.go
index 927da6b1..aab91337 100644
--- a/test/framework/alertmanager.go
+++ b/test/framework/alertmanager.go
@@ -18,6 +18,7 @@ import (
 	"encoding/json"
 	"fmt"
 	"os"
+	"strconv"
 	"strings"
 	"time"
 
@@ -126,19 +127,37 @@ func (f *Framework) CreateAlertmanagerAndWaitUntilReady(ns string, a *monitoring
 		return nil, errors.Wrap(err, fmt.Sprintf("creating alertmanager %v failed", a.Name))
 	}
 
-	return a, f.WaitForAlertmanagerReady(ns, a.Name, int(*a.Spec.Replicas))
+	return a, f.WaitForAlertmanagerClusterReady(ns, a.Name, int(*a.Spec.Replicas))
 }
 
-func (f *Framework) WaitForAlertmanagerReady(ns, name string, replicas int) error {
-	err := WaitForPodsReady(
+// WaitForAlertmanagerClusterReady waits for each individual pod as well as the
+// cluster as a whole to be ready.
+func (f *Framework) WaitForAlertmanagerClusterReady(ns, name string, replicas int) error {
+	if err := WaitForPodsReady(
 		f.KubeClient,
 		ns,
 		5*time.Minute,
 		replicas,
 		alertmanager.ListOptions(name),
-	)
+	); err != nil {
+		return errors.Wrap(err,
+			fmt.Sprintf(
+				"failed to wait for an Alertmanager cluster (%s) with %d instances to become ready",
+				name, replicas,
+			))
+	}
 
-	return errors.Wrap(err, fmt.Sprintf("failed to create an Alertmanager cluster (%s) with %d instances", name, replicas))
+	for i := 0; i < replicas; i++ {
+		name := fmt.Sprintf("alertmanager-%v-%v", name, strconv.Itoa(i))
+		if err := f.WaitForAlertmanagerInitializedMesh(ns, name, replicas); err != nil {
+			fmt.Sprintf(
+				"failed to wait for an Alertmanager cluster (%s) with %d instances to become ready",
+				name, replicas,
+			)
+		}
+	}
+
+	return nil
 }
 
 func (f *Framework) UpdateAlertmanagerAndWaitUntilReady(ns string, a *monitoringv1.Alertmanager) (*monitoringv1.Alertmanager, error) {
```

**Comment by paulfantom** [comment]
Looks good to me. Applied in latest commit.

**Comment by brancz** [comment]
lgtm

**Comment by paulfantom** [comment]
@mxinden seems like latest patch is making test timeout in travis (or not print anything for 10 minutes, so travis is killing this job). I have run it 4 times and it is a consistent behavior.

**Comment by mxinden** [comment]
> @mxinden seems like latest patch is making test timeout in travis (or not print anything for 10 minutes, so travis is killing this job). I have run it 4 times and it is a consistent behavior.

@paulfantom interesting. Looking into that.

**Comment by mxinden** [comment]
Last commit should remove the test timeout in travis.

**Comment by mxinden** [comment]
Imagine the scenario where two alertmanager clusters run in parallel.
Alertmanager cluster A drops instance with IP address 1. Alertmanager
cluster B adds a new instance which gets the recently freed IP adddress 1. 
If this happens in a reasonably short time, both clusters will merge.

The end-to-end test suit spins up many Alertmanager clusters in
parallel, scaling them up and down. The above scenario is happening
here.

For now we can run the tests in sequence. Long term Alertmanager should
have a cluster identifier to prevent such cluster merges.

**Comment by mxinden** [comment]
While running all Alertmanager tests in sequence increases the overall runtime of the tests (30 minutes total), it is a temporary solution to the above problem. What do you think @paulfantom? I would suggest squashing the commits and merging.

**Comment by paulfantom** [comment]
@mxinden sounds reasonable to me :+1:

**Comment by mxinden** [comment]
This seems to stabilize things. @paulfantom would you mind giving this another review? This is good to go from my side.

File: pkg/alertmanager/statefulset.go [hunk 1]
```diff
@@ -33,8 +33,10 @@ import (
 )
 
 const (
-	governingServiceName   = "alertmanager-operated"
-	defaultVersion         = "v0.16.1"
+	governingServiceName = "alertmanager-operated"
+	// DefaultVersion specifies which version of Alertmanager the Prometheus
+	// Operator uses by default.
+	DefaultVersion         = "v0.16.1"
```
Comment (mxinden) [inline pkg/alertmanager/statefulset.go]: Needed in `test/e2e`.

**Comment by brancz** [comment]
looks good :+1:

**Comment by brancz** [comment]
Thanks both of you for fixing this! It's been a journey! :tada:

**Comment by mxinden** [comment]
@brancz while this is a good step forward, we are still facing issues with mesh initialization on Alertmanager v0.14.0. So far I am unable to reproduce this locally. I will test out different Kubernetes versions next.

**Pull request merged**